hnn_core/optimization.py

```diff
@@ -0,0 +1,427 @@
+        evinput_params[input_name]['opt_end'] = ceil(
+            evinput_this['opt_end'] / dt) * dt
+
+    for evinput_this in evinput_params.values():
+        del evinput_this['mean'], evinput_this['sigma'], evinput_this['cdf']
+
+    return evinput_params
+
+
+def _create_last_chunk(input_chunks):
+    """ This creates a chunk that combines parameters for
+    all chunks in input_chunks (final step)
+
+    Parameters
+    ----------
+    input_chunks: List
+        List of ordered chunks for optimization
+
+    Returns
+    -------
+    chunk: dict
+        Dictionary of with parameters for combined
+        chunk (final step)
+    """
+    chunk = {'inputs': [], 'ranges': {}, 'opt_start': 0.0,
+             'opt_end': 0.0}
+
+    for evinput in input_chunks:
+        chunk['inputs'].extend(evinput['inputs'])
+        chunk['ranges'].update(evinput['ranges'])
+        if evinput['opt_end'] > chunk['opt_end']:
+            chunk['opt_end'] = evinput['opt_end']
+
+    # wRMSE with weights of 1's is the same as regular RMSE.
+    chunk['weights'] = np.ones_like(input_chunks[-1]['weights'])
+
+    return chunk
+
+
+def _consolidate_chunks(inputs):
+    """Consolidates inputs into optimization "chunks" defined by
+    opt_start and opt_end
+
+    Parameters
+    ----------
+    inputs: dict
+        Sorted dictionary of inputs with their parameters
+        and weight functions
+
+    Returns
+    -------
+    chunks: list
+        Combine the evinput_params whenever the end is overlapping
+        with the next.
+    """
+    chunks = list()
+    for input_name in inputs:
+        input_dict = inputs[input_name].copy()
+        input_dict['inputs'] = [input_name]
+
+        if (len(chunks) > 0 and
+                input_dict['start'] <= chunks[-1]['end']):
+            # update previous chunk
+            chunks[-1]['inputs'].extend(input_dict['inputs'])
+            chunks[-1]['end'] = input_dict['end']
+            chunks[-1]['ranges'].update(input_dict['ranges'])
+            chunks[-1]['opt_end'] = max(chunks[-1]['opt_end'],
+                                        input_dict['opt_end'])
+            # average the weights
+            chunks[-1]['weights'] = (chunks[-1]['weights'] +
+                                     input_dict['weights']) / 2
+        else:
+            # new chunk
+            chunks.append(input_dict)
+
+    # add one last chunk to the end
+    if len(chunks) > 1:
+        last_chunk = _create_last_chunk(chunks)
+        chunks.append(last_chunk)
+
+    return chunks
+
+
+def _optrun(new_params, opt_params, params, opt_dpls):
+    """ This is the function to run a simulation
+
+    Parameters
+    ----------
+    new_params: Array
+        List or numpy array with the parameters chosen by
+        optimization engine. Order is consistent with
+        opt_params['ranges'].
+    opt_params : dict
+        The optimization parameters
+    params : dict
+        The params dictionary.
```

Comment:
```suggestion
    params : an instance of Params
        Params containing paramter values from file
```
I think this applies to a couple more places (docstring pulled from `params.py`)